firebase:shell dumps function definition object to console

<!-- DO NOT DELETE 
validate_template=true
template_path=.github/ISSUE_TEMPLATE/bug_report.md
-->


### [REQUIRED] Environment info
firebase-tools: 7.1.1
firebase-admin: 8.2.0
firebase-functions: 3.1.0
node: 8.15.0
Linux

### [REQUIRED] Test case

Testing a function in the shell console cause a pretty big Object looking like a definition of all functions to be dumped in the console. This seems to be only happening after a threshold (either number of functions or volume of code). 

console.info messages are mixed into this dump: 
![image](https://user-images.githubusercontent.com/1168053/61542990-e51da700-aa42-11e9-97e4-2caf1731921b.png)


The Object looks like: 
```json

{
  "timestamp": "Fri Jul 19 2019 16:03:28 GMT+0200 (CEST)",
  "level": "SYSTEM",
  "text": "",
  "data":
  {
    "triggers":
    {
      "api":
      {
        "definition":
        {
          "httpsTrigger":
          {},
          "name": "api",
          "entryPoint": "api"
        },
        "module":
        {
          "appState":
          {},
          "arithmetic":
          {
            "counter":
            {},
            "average":
            {}
          },
          "delete":
          {},
          "entity":
          {
            "resource":
            {},
            "tenant":
            {},
            "program":
            {},
            "record":
            {},
            "business":
            {}
          },
          "resources":
          {
            "machine":
            {},
            "category":
            {}
          },
          "lowerCase":
          {},
          "membership":
          {},
          "user":
          {},
          "documentation":
          {},
          "schedule":
          {}
        }
      },
      "appState-userUpdateProgram":
      {
        "definition":
        {
          "eventTrigger":
          {
            "resource": "projects/_/instances/project-8476440599955222702/refs/userData/appState/{uid}/buid",
            "eventType": "providers/google.firebase.database/eventTypes/ref.write",
            "service": "firebaseio.com"
          },
          "name": "appState-userUpdateProgram",
          "entryPoint": "appState.userUpdateProgram"
        },
        "module":
        {
....
```

Thanks!

Almost
```
christophe ~/Programming/Polymer/preignition/functions $ sudo npm install -g git://github.com/firebase/firebase-tools.git#ss-ipc-logs --save
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.9 (node_modules/firebase-tools/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.9: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

npm ERR! path /usr/local/lib/node_modules/firebase-tools/lib/bin/firebase.js
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall chmod
npm ERR! enoent ENOENT: no such file or directory, chmod '/usr/local/lib/node_modules/firebase-tools/lib/bin/firebase.js'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/christophe/.npm/_logs/2019-07-19T19_14_48_933Z-debug.log
```
[2019-07-19T18_53_26_739Z-debug.log](https://github.com/firebase/firebase-tools/files/3412415/2019-07-19T18_53_26_739Z-debug.log)


I can install other npm modules (like firebase-tools) without problem

--- a/changelog.txt
+++ b/changelog.txt
@@ -1,2 +1,3 @@
 * Allow the RTDB emulator to hot-reload rules file on changes.
 * Allows emulated Cloud Functions to talk to production RTDB/Firestore if the emulators are not running.
+* Fixes an issue where internal logs would sometimes appear in stdout.

--- a/src/emulator/functionsEmulator.ts
+++ b/src/emulator/functionsEmulator.ts
@@ -585,6 +585,7 @@ You can probably fix this by running "npm install ${
       "functions",
       `Setting up Realtime Database trigger "${definition.name}"`
     );
+    logger.debug(`addDatabaseTrigger`, JSON.stringify(bundle));
     return new Promise((resolve, reject) => {
       let setTriggersPath = `http://localhost:${databasePort}/.settings/functionTriggers.json`;
       if (projectId !== "") {

--- a/src/emulator/functionsEmulator.ts
+++ b/src/emulator/functionsEmulator.ts
@@ -585,7 +585,6 @@ You can probably fix this by running "npm install ${
       "functions",
       `Setting up Realtime Database trigger "${definition.name}"`
     );
-    logger.debug(`addDatabaseTrigger`, JSON.stringify(bundle));
     return new Promise((resolve, reject) => {
       let setTriggersPath = `http://localhost:${databasePort}/.settings/functionTriggers.json`;
       if (projectId !== "") {
@@ -800,6 +799,11 @@ export function InvokeRuntime(
     stderr: { pipe: runtime.stderr, value: "" },
     stdout: { pipe: runtime.stdout, value: "" },
   };
+
+  const ipcBuffer = { value: "" };
+  runtime.on("message", (message: any) => {
+    onData(runtime, emitter, ipcBuffer, message);
+  });
 
   for (const id in buffers) {
     if (buffers.hasOwnProperty(id)) {

--- a/src/emulator/functionsEmulator.ts
+++ b/src/emulator/functionsEmulator.ts
@@ -841,18 +841,7 @@ function onData(
   buf: Buffer
 ): void {
   let bufString = buf.toString();
-
-  // Remove all chunk markings from the message
-  const endedWithChunk = bufString.endsWith(EmulatorLog.CHUNK_DIVIDER);
-  while (bufString.indexOf(EmulatorLog.CHUNK_DIVIDER) >= 0) {
-    bufString = bufString.replace(EmulatorLog.CHUNK_DIVIDER, "");
-  }
   buffer.value += bufString;
-
-  // If the message ended with a chunk divider, we just wait for more to come.
-  if (endedWithChunk) {
-    return;
-  }
 
   const lines = buffer.value.split("\n");
 

--- a/src/emulator/functionsEmulator.ts
+++ b/src/emulator/functionsEmulator.ts
@@ -840,8 +840,7 @@ function onData(
   buffer: { value: string },
   buf: Buffer
 ): void {
-  let bufString = buf.toString();
-  buffer.value += bufString;
+  buffer.value += buf.toString();
 
   const lines = buffer.value.split("\n");
 

--- a/src/emulator/types.ts
+++ b/src/emulator/types.ts
@@ -198,10 +198,23 @@ export class EmulatorLog {
     }
 
     EmulatorLog.WAITING_FOR_FLUSH = true;
-    process.stdout.write(nextMsg, () => {
-      EmulatorLog.WAITING_FOR_FLUSH = EmulatorLog.LOG_BUFFER.length > 0;
-      this.flush();
-    });
+    if (process.send) {
+      // For some reason our node.d.ts file does not include the version of subprocess.send() with a callback
+      // but the node docs assert that it has an optional callback.
+      // https://nodejs.org/api/child_process.html#child_process_subprocess_send_message_sendhandle_options_callback
+      (process.send as any)(nextMsg, undefined, {}, (err: any) => {
+        if (err) {
+          process.stderr.write(err);
+        }
+
+        EmulatorLog.WAITING_FOR_FLUSH = EmulatorLog.LOG_BUFFER.length > 0;
+        this.flush();
+      });
+    } else {
+      process.stderr.write(
+        "subprocess.send() is undefined, cannot communicate with Functions Runtime."
+      );
+    }
   }
 
   private chunkString(msg: string): string[] {

--- a/src/emulator/types.ts
+++ b/src/emulator/types.ts
@@ -76,10 +76,6 @@ export interface Address {
 }
 
 export class EmulatorLog {
-  // Messages over 8192 bytes cause issues
-  static CHUNK_DIVIDER = "__CHUNK__";
-  static CHUNK_SIZE = 8000;
-
   get date(): Date {
     if (!this.timestamp) {
       return new Date(0);
@@ -153,37 +149,13 @@ export class EmulatorLog {
   }
 
   /**
-   * As discovered in #1486, some very large log messages (>8192B) were not being properly passed
-   * between the emulator runtime and the emulator itself. We were falsely making the assumption
-   * that we could pass messages of any length over stdout and the stream reader would always get
-   * a whole message in a single "data" callback.
-   *
-   * Now we chunk the messages into 8000B pieces and then add a signal (CHUNK_DIVIDER) to the
-   * end of partial messages that instructs the receiver to wait for the whole message to
-   * appear.
-   *
    * We use a global boolean to know if all of our messages have been flushed, and the functions
    * emulator can wait on this variable to flip before exiting. This ensures that we never
-   * miss a log message that has been queued but has not yet flushed from stdout.
-   *
-   * Note: it would be better to use ipc via process.send() since that is faster and has
-   * extremely large message limits but in some experiments the IPC channel seemed to get
-   * full and not flush, so stdout remains.
+   * miss a log message that has been queued but has not yet flushed.
    */
   log(): void {
     const msg = `${this.toString()}\n`;
-    const chunks = this.chunkString(msg);
-
-    for (let i = 0; i < chunks.length; i++) {
-      const chunk = chunks[i];
-      const isLast = i === chunks.length - 1;
-      if (isLast) {
-        this.bufferMessage(chunk);
-      } else {
-        this.bufferMessage(chunk + EmulatorLog.CHUNK_DIVIDER);
-      }
-    }
-
+    this.bufferMessage(msg);
     this.flush();
   }
 
@@ -215,19 +187,6 @@ export class EmulatorLog {
         "subprocess.send() is undefined, cannot communicate with Functions Runtime."
       );
     }
-  }
-
-  private chunkString(msg: string): string[] {
-    const chunks: string[] = [];
-    const numChunks = Math.ceil(msg.length / EmulatorLog.CHUNK_SIZE);
-
-    for (let i = 0; i < numChunks; i++) {
-      const start = i * EmulatorLog.CHUNK_SIZE;
-      const length = EmulatorLog.CHUNK_SIZE;
-      chunks.push(msg.substr(start, length));
-    }
-
-    return chunks;
   }
 
   private toStringCore(pretty = false): string {

